mbus-api/src/receive.rs

```diff
@@ -0,0 +1,158 @@
+/// let raw_msg = &subscriber.next().await?;
+/// let msg: ReceivedMessage<RequestConfig, ReplyConfig> =
+///             raw_msg.try_into()?;
+///
+/// msg.respond(ReplyConfig {}).await.unwrap();
+/// // or we can also use the same fn to return an error
+/// msg.respond(Err(Error::Message("failure".into()))).await.unwrap();
+/// ```
+pub struct ReceivedMessage<'a, S, R> {
+    request: SendPayload<S>,
+    bus_message: &'a nats::asynk::Message,
+    reply_type: PhantomData<R>,
+}
+
+impl<'a, S, R> ReceivedMessage<'a, S, R>
+where
+    for<'de> S: Deserialize<'de> + 'a + Debug + Clone + MessageIdTrait,
+    R: Serialize,
+{
+    /// Get a copy of the actual payload data which was received.
+    pub fn payload(&self) -> S {
+        self.request.data.clone()
+    }
+    /// Get the sender identifier
+    pub fn sender(&self) -> SenderId {
+        self.request.sender.clone()
+    }
+
+    /// Respond back to the sender with the `reply` payload wrapped by
+    /// a Result-like type.
+    /// May fail if serialization of the reply fails or if the
+    /// message bus fails to respond.
+    /// Can receive either `R`, `Err()` or `ReplyPayload`.
+    pub async fn respond<T: Into<ReplyPayload<R>>>(
+        &self,
+        reply: T,
+    ) -> io::Result<()> {
+        let reply: ReplyPayload<R> = reply.into();
+        let payload = serde_json::to_vec(&reply)?;
+        self.bus_message.respond(&payload).await
+    }
+
+    /// Create a new received message object which wraps the send and
+    /// receive types around a raw bus message.
+    fn new(bus_msg: &'a nats::asynk::Message) -> Result<Self, io::Error> {
+        let request: SendPayload<S> = serde_json::from_slice(&bus_msg.data)?;
+        if request.id == request.data.id() {
+            log::info!(
+                "We have a message from '{}': {:?}",
+                request.sender,
+                request.data
+            );
+            Ok(Self {
+                request,
+                bus_message: bus_msg,
+                reply_type: Default::default(),
+            })
+        } else {
+            Err(io::Error::new(
+                io::ErrorKind::InvalidInput,
+                "invalid message id!",
+            ))
+        }
+    }
+}
+
+/// Message received over the message bus with a reply serialization wrapper
+/// For type safety refer to `ReceivedMessage<'a,S,R>`.
+pub struct ReceivedRawMessage<'a> {
+    bus_msg: &'a nats::asynk::Message,
+}
+
+impl<'a> ReceivedRawMessage<'a> {
+    /// Get a copy of the actual payload data which was sent
+    /// May fail if the raw data cannot be deserialized into `S`
+    pub fn payload<S: Deserialize<'a>>(&self) -> io::Result<S> {
+        let request: SendPayload<S> =
+            serde_json::from_slice(&self.bus_msg.data)?;
+        Ok(request.data)
+    }
+
+    /// Get the identifier of this message.
+    /// May fail if the raw data cannot be deserialized into the preamble.
+    pub fn id(&self) -> io::Result<MessageId> {
```

Comment:
why are we not using the messageID trait here?

Reply:
Because the MessageID trait is not fallible... though I guess we can make it fallible